controllers/nodes/nodes_controller.go

```diff
@@ -52,84 +62,81 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
-			&operatorhandler.InstrumentedEnqueueRequestForObject[client.Object]{},
+		source.Kind[*corev1.Node](
+			mgr.GetCache(), &corev1.Node{},
+			&operatorhandler.InstrumentedEnqueueRequestForObject[*corev1.Node]{},
 			nodeCountChangePredicate{},
 		))
 	if err != nil {
 		return err
 	}
 
-	return nil
-}
-
-// Custom predicate to detect changes in node count
-type nodeCountChangePredicate struct {
-	predicate.Funcs
-}
-
-func (nodeCountChangePredicate) Update(e event.UpdateEvent) bool {
-	return !maps.Equal(e.ObjectOld.GetLabels(), e.ObjectNew.GetLabels())
-}
-
-func (nodeCountChangePredicate) Create(_ event.CreateEvent) bool {
-	// node is added
-	return true
-}
-
-func (nodeCountChangePredicate) Delete(_ event.DeleteEvent) bool {
-	// node is removed
-	return true
-}
-
-func (nodeCountChangePredicate) Generic(_ event.GenericEvent) bool {
-	return false
+	return c.Watch(
+		source.Kind[*hcov1beta1.HyperConverged](
+			mgr.GetCache(), &hcov1beta1.HyperConverged{},
+			&handler.TypedEnqueueRequestForObject[*hcov1beta1.HyperConverged]{},
+			hyperconvergedPredicate{},
+		))
 }
 
 // ReconcileNodeCounter reconciles the nodes count
 type ReconcileNodeCounter struct {
 	// This client, initialized using mgr.Client() above, is a split client
 	// that reads objects from the cache and writes to the apiserver
-	client              client.Client
+	client.Client
 	HyperConvergedQueue workqueue.TypedRateLimitingInterface[reconcile.Request]
+	nodeEvents          chan<- event.GenericEvent
 }
 
 // Reconcile updates the nodes count on ClusterInfo singleton
-func (r *ReconcileNodeCounter) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
-	log.Info("Triggered by a node count change")
-
-	err := nodeinfo.HandleNodeChanges(ctx, r.client)
+func (r *ReconcileNodeCounter) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	logger, err := logr.FromContext(ctx)
 	if err != nil {
-		return reconcile.Result{}, err
+		logger = log
+	}
+	if req == hcoReq {
+		// This is a request triggered by a change in the HyperConverged CR
+		logger.Info("Triggered by a HyperConverged CR change")
+	} else {
+		logger.Info("Triggered by a node change", "node name", req.Name)
 	}
 
-	hco := &hcov1beta1.HyperConverged{}
-	namespace := hcoutil.GetOperatorNamespaceFromEnv()
-	hcoKey := types.NamespacedName{
-		Name:      hcoutil.HyperConvergedName,
-		Namespace: namespace,
+	logger.Info("Reading the latest HyperConverged CR")
+	hc, err := r.readHyperConverged(ctx)
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("failed to read the HyperConverged CR; %v", err)
 	}
-	err = r.client.Get(ctx, hcoKey, hco)
+
+	nodeInfoChanged, err := nodeinfo.HandleNodeChanges(ctx, r, hc, logger)
 	if err != nil {
-		if errors.IsNotFound(err) {
-			return reconcile.Result{}, nil
-		}
 		return reconcile.Result{}, err
 	}
 
-	if !hco.DeletionTimestamp.IsZero() {
+	if hc == nil || !hc.DeletionTimestamp.IsZero() {
```

Comment:
if `hc` is nil at that point, i think `nodeinfo.HandleNodeChanges(ctx, r, hc, logger)`  would return an error.

Reply:
hc == nil is handled. Look at the `isWorkloadNodeFunc` function in pkg/internal/nodeinfo/nodeinfo.go. This function choose the function to filter workloads nodes. so if hc is nil, it will return a function that only uses the node's worker label. And if it's not nil, it checks if hc's `spec.workloads.nodePlacement` field is populated.